src/java.base/share/classes/java/io/Reader.java

```diff
@@ -140,6 +140,124 @@ public void close() {
         };
     }
 
+    /**
+     * Returns a new {@code Reader} whose source is a {@link CharSequence}.
+     *
+     * <p> The resulting reader is not safe for use by multiple
+     * concurrent threads. If the reader is to be used by more than one
+     * thread it should be controlled by appropriate synchronization.
+     *
+     * <p> The returned reader is initially open. The reader is closed by
+     * calling the {@code close()} method. Subsequent calls to {@code close()}
+     * have no effect.
+     *
+     * <p> After the reader has been closed, the {@code read()},
+     * {@code read(char[])}, {@code read(char[], int, int)},
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The {@code markSupported()} method returns {@code true}.
+     *
+     * @param source {@code CharSequence} providing the character stream.
+     *
+     * @return a {@code Reader} which reads characters from {@code source}
+     *
+     * @throws NullPointerException if {@code source} is {@code null}
+     *
+     * @since 24
+     */
+    public static Reader of(CharSequence source) {
+        Objects.requireNonNull(source);
+
+        return new Reader() {
+            private final int length = source.length();
+            private CharSequence cs = source;
+            private int next = 0;
+            private int mark = 0;
+
+            /** Check to make sure that the stream has not been closed */
+            private void ensureOpen() throws IOException {
+                if (cs == null)
+                    throw new IOException("Stream closed");
+            }
+
+            @Override
+            public int read() throws IOException {
+                ensureOpen();
+                if (next >= length)
+                    return -1;
+                return cs.charAt(next++);
+            }
+
+            @Override
+            public int read(char[] cbuf, int off, int len) throws IOException {
+                ensureOpen();
+                Objects.checkFromIndexSize(off, len, cbuf.length);
+                if (len == 0) {
+                    return 0;
+                }
+                if (next >= length)
+                    return -1;
+                int n = Math.min(length - next, len);
+                switch (cs) {
+                    case String s -> s.getChars(next, next + n, cbuf, off);
```

Comment:
There was some discussion on the mailing list of introducing a method to CharSequence for bulk getChars. Doing that would help both here and in Appendable/Writer implementations like StringWriter, PrintWriter, and OutputStreamWriter which currently convert to a String to then write.

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/io/Writer.java#L367

Reply:
I believe this batch operation would be a useful addition, but it is best reserved for another RFE.

Reply:
I just filed the PR to implement `CharSequence.getChars` bulk-read method. Please see https://github.com/openjdk/jdk/pull/21730. 🙂